<!DOCTYPE html><html lang="en-US" prefix="og: https://ogp.me/ns#"><!-- Mirrored from developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/try by HTTrack Website Copier/3.x [XR&CO'2014], Wed, 08 Jan 2025 13:13:27 GMT --><head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1"><link rel="icon" href="https://developer.mozilla.org/favicon-48x48.bc390275e955dacb2e65.png"><link rel="apple-touch-icon" href="https://developer.mozilla.org/apple-touch-icon.528534bba673c38049c2.png"><meta name="theme-color" content="#ffffff"><link rel="manifest" href="https://developer.mozilla.org/manifest.f42880861b394dd4dc9b.json"><link rel="search" type="application/opensearchdescription+xml" href="https://developer.mozilla.org/opensearch.xml" title="MDN Web Docs"><title>Promise.try() - JavaScript | MDN</title><link rel="alternate" title="Promise.try()" href="https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Promise/try" hreflang="de"><link rel="alternate" title="Promise.try()" href="https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Promise/try" hreflang="ja"><link rel="alternate" title="Promise.try()" href="https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/Promise/try" hreflang="ko"><link rel="alternate" title="Promise.try()" href="https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Promise/try" hreflang="zh"><link rel="alternate" title="Promise.try()" href="try.html" hreflang="en"><link rel="preload" as="font" type="font/woff2" href="https://developer.mozilla.org/static/media/Inter.var.c2fe3cb2b7c746f7966a.woff2" crossorigin=""><link rel="alternate" type="application/rss+xml" title="MDN Blog RSS Feed" href="https://developer.mozilla.org/en-US/blog/rss.xml" hreflang="en"><meta name="description" content="The Promise.try() static method takes a callback of any kind (returns or throws, synchronously or asynchronously) and wraps its result in a Promise."><meta property="og:url" content="try.html"><meta property="og:title" content="Promise.try() - JavaScript | MDN"><meta property="og:type" content="website"><meta property="og:locale" content="en_US"><meta property="og:description" content="The Promise.try() static method takes a callback of any kind (returns or throws, synchronously or asynchronously) and wraps its result in a Promise."><meta property="og:image" content="https://developer.mozilla.org/mdn-social-share.d893525a4fb5fb1f67a2.png"><meta property="og:image:type" content="image/png"><meta property="og:image:height" content="1080"><meta property="og:image:width" content="1920"><meta property="og:image:alt" content="The MDN Web Docs logo, featuring a blue accent color, displayed on a solid black background."><meta property="og:site_name" content="MDN Web Docs"><meta name="twitter:card" content="summary_large_image"><meta name="twitter:creator" content="MozDevNet"><link rel="canonical" href="try.html"><style media="print">.article-actions-container,.document-toc-container,.language-menu,.main-menu-toggle,.on-github,.page-footer,.place,.sidebar,.top-banner,.top-navigation-main,ul.prev-next{display:none!important}.main-page-content,.main-page-content pre{padding:2px}.main-page-content pre{border-left-width:2px}</style><link href="https://developer.mozilla.org/static/css/main.65aa8eed.css" rel="stylesheet"></head><body><script>if(document.body.addEventListener("load",(t=>{t.target.classList.contains("interactive")&&t.target.setAttribute("data-readystate","complete")}),{capture:!0}),window&&document.documentElement){const t={light:"#ffffff",dark:"#1b1b1b"};try{const e=window.localStorage.getItem("theme");e&&(document.documentElement.className=e,document.documentElement.style.backgroundColor=t[e]);const o=window.localStorage.getItem("nop");o&&(document.documentElement.dataset.nop=o)}catch(t){console.warn("Unable to read theme from localStorage",t)}}</script><div id="root"><div class="page-wrapper  category-javascript document-page"><div class="main-wrapper"><main id="content" class="main-content  "><article class="main-page-content" lang="en-US"><header><h1>Promise.try()</h1><details class="baseline-indicator not"><summary><span class="indicator" role="img" aria-label="Baseline Cross"></span><h2><span class="not-bold">Limited availability</span></h2><div class="browsers"><span class="engine" title="Supported in Chrome and Edge"><span class="browser chrome supported" role="img" aria-label="Chrome check"></span><span class="browser edge supported" role="img" aria-label="Edge check"></span></span><span class="engine" title="Not widely supported in Firefox"><span class="browser firefox " role="img" aria-label="Firefox cross"></span></span><span class="engine" title="Supported in Safari"><span class="browser safari supported" role="img" aria-label="Safari check"></span></span></div><span class="icon icon-chevron "></span></summary><div class="extra"><p>This feature is not Baseline because it does not work in some of the most widely-used browsers.</p><ul><li><a href="https://developer.mozilla.org/en-US/docs/Glossary/Baseline/Compatibility" data-glean="baseline_link_learn_more" target="_blank" class="learn-more">Learn more</a></li><li><a href="#browser_compatibility" data-glean="baseline_link_bcd_table">See full compatibility</a></li><li><a href="https://survey.alchemer.com/s3/7634825/MDN-baseline-feedback?page=%2Fen-US%2Fdocs%2FWeb%2FJavaScript%2FReference%2FGlobal_Objects%2FPromise%2Ftry&amp;level=not" data-glean="baseline_link_feedback" class="feedback-link" target="_blank" rel="noreferrer">Report feedback</a></li></ul></div></details></header><div class="section-content"><p>The <strong><code>Promise.try()</code></strong> static method takes a callback of any kind (returns or throws, synchronously or asynchronously) and wraps its result in a <a href="../Promise.html"><code>Promise</code></a>.</p></div><section aria-labelledby="syntax"><h2 id="syntax"><a href="#syntax">Syntax</a></h2><div class="section-content"><div class="code-example"><div class="example-header"><span class="language-name">js</span></div><pre class="brush: js notranslate"><code>Promise.try(func)
Promise.try(func, arg1)
Promise.try(func, arg1, arg2)
Promise.try(func, arg1, arg2, /* …, */ argN)
</code></pre></div></div></section><section aria-labelledby="parameters"><h3 id="parameters"><a href="#parameters">Parameters</a></h3><div class="section-content"><dl>
<dt id="func"><a href="#func"><code>func</code></a></dt>
<dd>
<p>A function that is called synchronously with the arguments provided (<code>arg1</code>, <code>arg2</code>, …, <code>argN</code>). It can do anything—either return a value, throw an error, or return a promise.</p>
</dd>
<dt id="arg1"><a href="#arg1"><code>arg1</code>, <code>arg2</code>, …, <code>argN</code></a></dt>
<dd>
<p>Arguments to pass to <code>func</code>.</p>
</dd>
</dl></div></section><section aria-labelledby="return_value"><h3 id="return_value"><a href="#return_value">Return value</a></h3><div class="section-content"><p>A <a href="../Promise.html"><code>Promise</code></a> that is:</p>
<ul>
<li>Already fulfilled, if <code>func</code> synchronously returns a value.</li>
<li>Already rejected, if <code>func</code> synchronously throws an error.</li>
<li>Asynchronously fulfilled or rejected, if <code>func</code> returns a promise.</li>
</ul></div></section><section aria-labelledby="description"><h2 id="description"><a href="#description">Description</a></h2><div class="section-content"><p>You may have an API that takes a callback. The callback may be synchronous or asynchronous. You want to handle everything uniformly by wrapping the result in a promise. The most straightforward way might be <a href="resolve.html"><code>Promise.resolve(func())</code></a>. The problem is that if <code>func()</code> synchronously throws an error, this error would not be caught and turned into a rejected promise.</p>
<p>The common approach (lifting a function call result into a promise, fulfilled or rejected) often looks like this:</p>
<div class="code-example"><div class="example-header"><span class="language-name">js</span></div><pre class="brush: js notranslate"><code>new Promise((resolve) =&gt; resolve(func()));
</code></pre></div>
<p>But <code>Promise.try()</code> is more helpful here:</p>
<div class="code-example"><div class="example-header"><span class="language-name">js</span></div><pre class="brush: js notranslate"><code>Promise.try(func);
</code></pre></div>
<p>For the built-in <code>Promise()</code> constructor, errors thrown from the executor are automatically caught and turned into rejections, so these two approaches are mostly equivalent, except that <code>Promise.try()</code> is more concise and readable.</p>
<p>Note that <code>Promise.try()</code> is <em>not</em> equivalent to this, despite being highly similar:</p>
<div class="code-example"><div class="example-header"><span class="language-name">js</span></div><pre class="brush: js notranslate"><code>Promise.resolve().then(func);
</code></pre></div>
<p>The difference is that the callback passed to <a href="then.html"><code>then()</code></a> is always called asynchronously, while the executor of the <code>Promise()</code> constructor is called synchronously. <code>Promise.try</code> also calls the function synchronously, and resolves the promise immediately if possible.</p>
<p><code>Promise.try()</code>, combined with <a href="catch.html"><code>catch()</code></a> and <a href="finally.html"><code>finally()</code></a>, can be used to handle both synchronous and asynchronous errors in a single chain, and make promise error handling appear almost like synchronous error handling.</p>
<p>Like <a href="https://developer.mozilla.org/en-US/docs/Web/API/Window/setTimeout" title="setTimeout()"><code>setTimeout()</code></a>, <code>Promise.try()</code> accepts extra arguments that are passed to the callback. This means instead of doing this:</p>
<div class="code-example"><div class="example-header"><span class="language-name">js</span></div><pre class="brush: js notranslate"><code>Promise.try(() =&gt; func(arg1, arg2));
</code></pre></div>
<p>You can do this:</p>
<div class="code-example"><div class="example-header"><span class="language-name">js</span></div><pre class="brush: js notranslate"><code>Promise.try(func, arg1, arg2);
</code></pre></div>
<p>Which are equivalent, but the latter avoids creating an extra closure and is more efficient.</p></div></section><section aria-labelledby="examples"><h2 id="examples"><a href="#examples">Examples</a></h2><div class="section-content"></div></section><section aria-labelledby="using_promise.try"><h3 id="using_promise.try"><a href="#using_promise.try">Using Promise.try()</a></h3><div class="section-content"><p>The following example takes a callback, "lifts" it into a promise, handles the result, and does some error handling:</p>
<div class="code-example"><div class="example-header"><span class="language-name">js</span></div><pre class="brush: js notranslate"><code>function doSomething(action) {
  return Promise.try(action)
    .then((result) =&gt; console.log(result))
    .catch((error) =&gt; console.error(error))
    .finally(() =&gt; console.log("Done"));
}

doSomething(() =&gt; "Sync result");

doSomething(() =&gt; {
  throw new Error("Sync error");
});

doSomething(async () =&gt; "Async result");

doSomething(async () =&gt; {
  throw new Error("Async error");
});
</code></pre></div>
<p>In async/await, the same code would look like this:</p>
<div class="code-example"><div class="example-header"><span class="language-name">js</span></div><pre class="brush: js notranslate"><code>async function doSomething(action) {
  try {
    const result = await action();
    console.log(result);
  } catch (error) {
    console.error(error);
  } finally {
    console.log("Done");
  }
}
</code></pre></div></div></section><section aria-labelledby="calling_try_on_a_non-promise_constructor"><h3 id="calling_try_on_a_non-promise_constructor"><a href="#calling_try_on_a_non-promise_constructor">Calling try() on a non-Promise constructor</a></h3><div class="section-content"><p><code>Promise.try()</code> is a generic method. It can be called on any constructor that implements the same signature as the <code>Promise()</code> constructor.</p>
<p>The following is a slightly more faithful approximation of the actual <code>Promise.try()</code> (although it should still not be used as a polyfill):</p>
<div class="code-example"><div class="example-header"><span class="language-name">js</span></div><pre class="brush: js notranslate"><code>Promise.try = function (func) {
  return new this((resolve, reject) =&gt; {
    try {
      resolve(func());
    } catch (error) {
      reject(error);
    }
  });
};
</code></pre></div>
<p>Because of how <code>Promise.try()</code> is implemented (i.e., the <code>try...catch</code>), we can safely invoke <code>Promise.try()</code> with its <code>this</code> set to any custom constructor, and it will never synchronously throw an error.</p>
<div class="code-example"><div class="example-header"><span class="language-name">js</span></div><pre class="brush: js notranslate"><code>class NotPromise {
  constructor(executor) {
    // The "resolve" and "reject" functions behave nothing like the native
    // promise's, but Promise.try() just calls resolve
    executor(
      (value) =&gt; console.log("Resolved", value),
      (reason) =&gt; console.log("Rejected", reason),
    );
  }
}

const p = Promise.try.call(NotPromise, () =&gt; "hello");
// Logs: Resolved hello

const p2 = Promise.try.call(NotPromise, () =&gt; {
  throw new Error("oops");
});
// Logs: Rejected Error: oops
</code></pre></div>
<p>Unlike <code>Promise()</code>, this <code>NotPromise()</code> constructor <em>does not</em> gracefully handle exceptions while running the executor. But despite the <code>throw</code>, <code>Promise.try()</code> still catches the exception, passing it to <code>reject()</code> to log out.</p></div></section><h2 id="specifications"><a href="#specifications">Specifications</a></h2><table class="standard-table"><thead><tr><th scope="col">Specification</th></tr></thead><tbody><tr><td><a href="https://tc39.es/proposal-promise-try/#sec-promise.try">Promise.try<!-- --> <br><small># <!-- -->sec-promise.try</small></a></td></tr></tbody></table><h2 id="browser_compatibility"><a href="#browser_compatibility">Browser compatibility</a></h2><p>BCD tables only load in the browser<noscript> <!-- -->with JavaScript enabled. Enable JavaScript to view data.</noscript></p><section aria-labelledby="see_also"><h2 id="see_also"><a href="#see_also">See also</a></h2><div class="section-content"><ul>
<li><a href="https://github.com/zloirock/core-js#promisetry" class="external" target="_blank">Polyfill of <code>Promise.try</code> in <code>core-js</code></a></li>
<li><a href="../../../Guide/Using_promises.html">Using promises</a> guide</li>
<li><a href="../Promise.html"><code>Promise</code></a></li>
<li><a href="Promise.html"><code>Promise()</code> constructor</a></li>
</ul></div></section></article></main></div></div></div><script type="application/json" id="hydration">{"url":"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/try","doc":{"body":[{"type":"prose","value":{"id":null,"title":null,"isH3":false,"content":"<p>The <strong><code>Promise.try()</code></strong> static method takes a callback of any kind (returns or throws, synchronously or asynchronously) and wraps its result in a <a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise\"><code>Promise</code></a>.</p>"}},{"type":"prose","value":{"id":"syntax","title":"Syntax","isH3":false,"content":"<div class=\"code-example\"><div class=\"example-header\"><span class=\"language-name\">js</span></div><pre class=\"brush: js notranslate\"><code>Promise.try(func)\nPromise.try(func, arg1)\nPromise.try(func, arg1, arg2)\nPromise.try(func, arg1, arg2, /* …, */ argN)\n</code></pre></div>"}},{"type":"prose","value":{"id":"parameters","title":"Parameters","isH3":true,"content":"<dl>\n<dt id=\"func\"><a href=\"#func\"><code>func</code></a></dt>\n<dd>\n<p>A function that is called synchronously with the arguments provided (<code>arg1</code>, <code>arg2</code>, …, <code>argN</code>). It can do anything—either return a value, throw an error, or return a promise.</p>\n</dd>\n<dt id=\"arg1\"><a href=\"#arg1\"><code>arg1</code>, <code>arg2</code>, …, <code>argN</code></a></dt>\n<dd>\n<p>Arguments to pass to <code>func</code>.</p>\n</dd>\n</dl>"}},{"type":"prose","value":{"id":"return_value","title":"Return value","isH3":true,"content":"<p>A <a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise\"><code>Promise</code></a> that is:</p>\n<ul>\n<li>Already fulfilled, if <code>func</code> synchronously returns a value.</li>\n<li>Already rejected, if <code>func</code> synchronously throws an error.</li>\n<li>Asynchronously fulfilled or rejected, if <code>func</code> returns a promise.</li>\n</ul>"}},{"type":"prose","value":{"id":"description","title":"Description","isH3":false,"content":"<p>You may have an API that takes a callback. The callback may be synchronous or asynchronous. You want to handle everything uniformly by wrapping the result in a promise. The most straightforward way might be <a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve\"><code>Promise.resolve(func())</code></a>. The problem is that if <code>func()</code> synchronously throws an error, this error would not be caught and turned into a rejected promise.</p>\n<p>The common approach (lifting a function call result into a promise, fulfilled or rejected) often looks like this:</p>\n<div class=\"code-example\"><div class=\"example-header\"><span class=\"language-name\">js</span></div><pre class=\"brush: js notranslate\"><code>new Promise((resolve) =&gt; resolve(func()));\n</code></pre></div>\n<p>But <code>Promise.try()</code> is more helpful here:</p>\n<div class=\"code-example\"><div class=\"example-header\"><span class=\"language-name\">js</span></div><pre class=\"brush: js notranslate\"><code>Promise.try(func);\n</code></pre></div>\n<p>For the built-in <code>Promise()</code> constructor, errors thrown from the executor are automatically caught and turned into rejections, so these two approaches are mostly equivalent, except that <code>Promise.try()</code> is more concise and readable.</p>\n<p>Note that <code>Promise.try()</code> is <em>not</em> equivalent to this, despite being highly similar:</p>\n<div class=\"code-example\"><div class=\"example-header\"><span class=\"language-name\">js</span></div><pre class=\"brush: js notranslate\"><code>Promise.resolve().then(func);\n</code></pre></div>\n<p>The difference is that the callback passed to <a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then\"><code>then()</code></a> is always called asynchronously, while the executor of the <code>Promise()</code> constructor is called synchronously. <code>Promise.try</code> also calls the function synchronously, and resolves the promise immediately if possible.</p>\n<p><code>Promise.try()</code>, combined with <a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch\"><code>catch()</code></a> and <a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally\"><code>finally()</code></a>, can be used to handle both synchronous and asynchronous errors in a single chain, and make promise error handling appear almost like synchronous error handling.</p>\n<p>Like <a href=\"/en-US/docs/Web/API/Window/setTimeout\" title=\"setTimeout()\"><code>setTimeout()</code></a>, <code>Promise.try()</code> accepts extra arguments that are passed to the callback. This means instead of doing this:</p>\n<div class=\"code-example\"><div class=\"example-header\"><span class=\"language-name\">js</span></div><pre class=\"brush: js notranslate\"><code>Promise.try(() =&gt; func(arg1, arg2));\n</code></pre></div>\n<p>You can do this:</p>\n<div class=\"code-example\"><div class=\"example-header\"><span class=\"language-name\">js</span></div><pre class=\"brush: js notranslate\"><code>Promise.try(func, arg1, arg2);\n</code></pre></div>\n<p>Which are equivalent, but the latter avoids creating an extra closure and is more efficient.</p>"}},{"type":"prose","value":{"id":"examples","title":"Examples","isH3":false,"content":""}},{"type":"prose","value":{"id":"using_promise.try","title":"Using Promise.try()","isH3":true,"content":"<p>The following example takes a callback, \"lifts\" it into a promise, handles the result, and does some error handling:</p>\n<div class=\"code-example\"><div class=\"example-header\"><span class=\"language-name\">js</span></div><pre class=\"brush: js notranslate\"><code>function doSomething(action) {\n  return Promise.try(action)\n    .then((result) =&gt; console.log(result))\n    .catch((error) =&gt; console.error(error))\n    .finally(() =&gt; console.log(\"Done\"));\n}\n\ndoSomething(() =&gt; \"Sync result\");\n\ndoSomething(() =&gt; {\n  throw new Error(\"Sync error\");\n});\n\ndoSomething(async () =&gt; \"Async result\");\n\ndoSomething(async () =&gt; {\n  throw new Error(\"Async error\");\n});\n</code></pre></div>\n<p>In async/await, the same code would look like this:</p>\n<div class=\"code-example\"><div class=\"example-header\"><span class=\"language-name\">js</span></div><pre class=\"brush: js notranslate\"><code>async function doSomething(action) {\n  try {\n    const result = await action();\n    console.log(result);\n  } catch (error) {\n    console.error(error);\n  } finally {\n    console.log(\"Done\");\n  }\n}\n</code></pre></div>"}},{"type":"prose","value":{"id":"calling_try_on_a_non-promise_constructor","title":"Calling try() on a non-Promise constructor","isH3":true,"content":"<p><code>Promise.try()</code> is a generic method. It can be called on any constructor that implements the same signature as the <code>Promise()</code> constructor.</p>\n<p>The following is a slightly more faithful approximation of the actual <code>Promise.try()</code> (although it should still not be used as a polyfill):</p>\n<div class=\"code-example\"><div class=\"example-header\"><span class=\"language-name\">js</span></div><pre class=\"brush: js notranslate\"><code>Promise.try = function (func) {\n  return new this((resolve, reject) =&gt; {\n    try {\n      resolve(func());\n    } catch (error) {\n      reject(error);\n    }\n  });\n};\n</code></pre></div>\n<p>Because of how <code>Promise.try()</code> is implemented (i.e., the <code>try...catch</code>), we can safely invoke <code>Promise.try()</code> with its <code>this</code> set to any custom constructor, and it will never synchronously throw an error.</p>\n<div class=\"code-example\"><div class=\"example-header\"><span class=\"language-name\">js</span></div><pre class=\"brush: js notranslate\"><code>class NotPromise {\n  constructor(executor) {\n    // The \"resolve\" and \"reject\" functions behave nothing like the native\n    // promise's, but Promise.try() just calls resolve\n    executor(\n      (value) =&gt; console.log(\"Resolved\", value),\n      (reason) =&gt; console.log(\"Rejected\", reason),\n    );\n  }\n}\n\nconst p = Promise.try.call(NotPromise, () =&gt; \"hello\");\n// Logs: Resolved hello\n\nconst p2 = Promise.try.call(NotPromise, () =&gt; {\n  throw new Error(\"oops\");\n});\n// Logs: Rejected Error: oops\n</code></pre></div>\n<p>Unlike <code>Promise()</code>, this <code>NotPromise()</code> constructor <em>does not</em> gracefully handle exceptions while running the executor. But despite the <code>throw</code>, <code>Promise.try()</code> still catches the exception, passing it to <code>reject()</code> to log out.</p>"}},{"type":"specifications","value":{"id":"specifications","title":"Specifications","isH3":false,"specifications":[{"bcdSpecificationURL":"https://tc39.es/proposal-promise-try/#sec-promise.try","title":"Promise.try"}],"query":"javascript.builtins.Promise.try"}},{"type":"browser_compatibility","value":{"id":"browser_compatibility","title":"Browser compatibility","isH3":false,"query":"javascript.builtins.Promise.try"}},{"type":"prose","value":{"id":"see_also","title":"See also","isH3":false,"content":"<ul>\n<li><a href=\"https://github.com/zloirock/core-js#promisetry\" class=\"external\" target=\"_blank\">Polyfill of <code>Promise.try</code> in <code>core-js</code></a></li>\n<li><a href=\"/en-US/docs/Web/JavaScript/Guide/Using_promises\">Using promises</a> guide</li>\n<li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise\"><code>Promise</code></a></li>\n<li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise\"><code>Promise()</code> constructor</a></li>\n</ul>"}}],"isActive":true,"isMarkdown":true,"isTranslated":false,"locale":"en-US","mdn_url":"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/try","modified":"2024-11-07T15:33:32.000Z","native":"English (US)","noIndexing":false,"other_translations":[{"locale":"de","title":"Promise.try()","native":"Deutsch"},{"locale":"ja","title":"Promise.try()","native":"日本語"},{"locale":"ko","title":"Promise.try()","native":"한국어"},{"locale":"zh-CN","title":"Promise.try()","native":"中文 (简体)"}],"pageTitle":"Promise.try() - JavaScript | MDN","parents":[{"uri":"/en-US/docs/Web","title":"References"},{"uri":"/en-US/docs/Web/JavaScript","title":"JavaScript"},{"uri":"/en-US/docs/Web/JavaScript/Reference","title":"Reference"},{"uri":"/en-US/docs/Web/JavaScript/Reference/Global_Objects","title":"Standard built-in objects"},{"uri":"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise","title":"Promise"},{"uri":"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/try","title":"Promise.try()"}],"popularity":null,"short_title":"Promise.try()","sidebarHTML":"<ol><li class=\"section\"><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects\">Standard built-in objects</a></li><li class=\"section\"><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise\"><code>Promise</code></a></li><li class=\"toggle\"><details open=\"\"><summary>Constructor</summary><ol><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise\"><code>Promise() constructor</code></a></li></ol></details></li><li class=\"toggle\"><details open=\"\"><summary>Static methods</summary><ol><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all\"><code>Promise.all()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled\"><code>Promise.allSettled()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any\"><code>Promise.any()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race\"><code>Promise.race()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/reject\"><code>Promise.reject()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve\"><code>Promise.resolve()</code></a></li><li><em><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/try\" aria-current=\"page\"><code>Promise.try()</code></a></em></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers\"><code>Promise.withResolvers()</code></a></li></ol></details></li><li class=\"toggle\"><details open=\"\"><summary>Static properties</summary><ol><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Symbol.species\"><code>Promise[Symbol.species]</code></a></li></ol></details></li><li class=\"toggle\"><details open=\"\"><summary>Instance methods</summary><ol><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch\"><code>Promise.prototype.catch()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally\"><code>Promise.prototype.finally()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then\"><code>Promise.prototype.then()</code></a></li></ol></details></li><li class=\"section\">Inheritance</li><li class=\"section\"><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object\"><code>Object/Function</code></a></li><li class=\"toggle\"><details><summary>Static methods</summary><ol><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply\"><code>Function.prototype.apply()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind\"><code>Function.prototype.bind()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/call\"><code>Function.prototype.call()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/toString\"><code>Function.prototype.toString()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/Symbol.hasInstance\"><code>Function.prototype[Symbol.hasInstance]()</code></a></li></ol></details></li><li class=\"toggle\"><details><summary>Static properties</summary><ol><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/displayName\"><code>Function: displayName</code></a><abbr class=\"icon icon-nonstandard\" title=\"Non-standard. Check cross-browser support before using.\">\n<span class=\"visually-hidden\">Non-standard</span>\n</abbr></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/length\"><code>Function: length</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/name\"><code>Function: name</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype\"><code>Function: prototype</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/arguments\"><code>Function.prototype.arguments</code></a><abbr class=\"icon icon-nonstandard\" title=\"Non-standard. Check cross-browser support before using.\">\n<span class=\"visually-hidden\">Non-standard</span>\n</abbr><abbr class=\"icon icon-deprecated\" title=\"Deprecated. Not for use in new websites.\">\n<span class=\"visually-hidden\">Deprecated</span>\n</abbr></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/caller\"><code>Function.prototype.caller</code></a><abbr class=\"icon icon-nonstandard\" title=\"Non-standard. Check cross-browser support before using.\">\n<span class=\"visually-hidden\">Non-standard</span>\n</abbr><abbr class=\"icon icon-deprecated\" title=\"Deprecated. Not for use in new websites.\">\n<span class=\"visually-hidden\">Deprecated</span>\n</abbr></li></ol></details></li><li class=\"toggle\"><details><summary>Instance methods</summary><ol><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/__defineGetter__\"><code>Object.prototype.__defineGetter__()</code></a><abbr class=\"icon icon-deprecated\" title=\"Deprecated. Not for use in new websites.\">\n<span class=\"visually-hidden\">Deprecated</span>\n</abbr></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/__defineSetter__\"><code>Object.prototype.__defineSetter__()</code></a><abbr class=\"icon icon-deprecated\" title=\"Deprecated. Not for use in new websites.\">\n<span class=\"visually-hidden\">Deprecated</span>\n</abbr></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/__lookupGetter__\"><code>Object.prototype.__lookupGetter__()</code></a><abbr class=\"icon icon-deprecated\" title=\"Deprecated. Not for use in new websites.\">\n<span class=\"visually-hidden\">Deprecated</span>\n</abbr></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/__lookupSetter__\"><code>Object.prototype.__lookupSetter__()</code></a><abbr class=\"icon icon-deprecated\" title=\"Deprecated. Not for use in new websites.\">\n<span class=\"visually-hidden\">Deprecated</span>\n</abbr></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty\"><code>Object.prototype.hasOwnProperty()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/isPrototypeOf\"><code>Object.prototype.isPrototypeOf()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/propertyIsEnumerable\"><code>Object.prototype.propertyIsEnumerable()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toLocaleString\"><code>Object.prototype.toLocaleString()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toString\"><code>Object.prototype.toString()</code></a></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/valueOf\"><code>Object.prototype.valueOf()</code></a></li></ol></details></li><li class=\"toggle\"><details><summary>Instance properties</summary><ol><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/proto\"><code>Object.prototype.__proto__</code></a><abbr class=\"icon icon-deprecated\" title=\"Deprecated. Not for use in new websites.\">\n<span class=\"visually-hidden\">Deprecated</span>\n</abbr></li><li><a href=\"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/constructor\"><code>Object.prototype.constructor</code></a></li></ol></details></li></ol>","source":{"folder":"en-us/web/javascript/reference/global_objects/promise/try","github_url":"https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/global_objects/promise/try/index.md","last_commit_url":"https://github.com/mdn/content/commit/7cd4706990ab95794415aee05ba0a9662e742a17","filename":"index.md"},"summary":"The Promise.try() static method takes a callback of any kind (returns or throws, synchronously or asynchronously) and wraps its result in a Promise.","title":"Promise.try()","toc":[{"text":"Syntax","id":"syntax"},{"text":"Description","id":"description"},{"text":"Examples","id":"examples"},{"text":"Specifications","id":"specifications"},{"text":"Browser compatibility","id":"browser_compatibility"},{"text":"See also","id":"see_also"}],"baseline":{"baseline":false,"support":{"chrome":"128","chrome_android":"128","edge":"128","safari":"18.2","safari_ios":"18.2"}},"browserCompat":["javascript.builtins.Promise.try"],"pageType":"javascript-static-method"}}</script>

</body><!-- Mirrored from developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/try by HTTrack Website Copier/3.x [XR&CO'2014], Wed, 08 Jan 2025 13:13:27 GMT --></html>